plugins/builds/helper/updateBuild.js

```diff
@@ -0,0 +1,304 @@
+ *
+ * @method isFixedBuild
+ * @param  build         Build Object
+ * @param  jobFactory    Job Factory instance
+ */
+async function isFixedBuild(build, jobFactory) {
+    if (build.status !== 'SUCCESS') {
+        return false;
+    }
+
+    const job = await jobFactory.get(build.jobId);
+    const failureBuild = await job.getLatestBuild({ status: 'FAILURE' });
+    const successBuild = await job.getLatestBuild({ status: 'SUCCESS' });
+
+    if ((failureBuild && !successBuild) || failureBuild.id > successBuild.id) {
+        return true;
+    }
+
+    return false;
+}
+
+/**
+ * Stops a frozen build from executing
+ * @method stopFrozenBuild
+ * @param  {Object} build         Build Object
+ * @param  {String} previousStatus    Prevous build status
+ */
+async function stopFrozenBuild(build, previousStatus) {
+    if (previousStatus !== 'FROZEN') {
+        return Promise.resolve();
+    }
+
+    return build.stopFrozen(previousStatus);
+}
+
+/**
+ * Updates execution details for init step
+ * @method stopFrozenBuild
+ * @param  {Object} build       Build Object
+ * @param  {Object} app         Hapi app Object
+ */
+async function updateInitStep(build, app) {
+    const step = await app.stepFactory.get({ buildId: build.id, name: 'sd-setup-init' });
+    // If there is no init step, do nothing
+
+    if (!step) {
+        return null;
+    }
+
+    step.endTime = build.startTime || new Date().toISOString();
+    step.code = 0;
+
+    return step.update();
+}
+
+/**
+ * Set build status to desired status, set build statusMessage
+ * @param {Object} build                Build Model
+ * @param {String} desiredStatus        New Status
+ * @param {String} statusMessage        User passed status message
+ * @param {String} statusMessageType    User passed severity of the status message
+ * @param {String} username             User initiating status build update
+ */
+function updateBuildStatus(build, desiredStatus, statusMessage, statusMessageType, username) {
+    // UNSTABLE -> SUCCESS needs to update meta and endtime.
+    // However, the status itself cannot be updated to SUCCESS
+    const currentStatus = build.status;
+
+    if (currentStatus !== 'UNSTABLE') {
+        if (desiredStatus !== undefined) {
+            build.status = desiredStatus;
+        }
+        if (build.status === 'ABORTED') {
+            if (currentStatus === 'FROZEN') {
+                build.statusMessage = `Frozen build aborted by ${username}`;
+            } else {
+                build.statusMessage = `Aborted by ${username}`;
+            }
+        } else if (build.status === 'FAILURE' || build.status === 'SUCCESS') {
+            if (statusMessage) {
+                build.statusMessage = statusMessage;
+                build.statusMessageType = statusMessageType || null;
+            }
+        } else {
+            build.statusMessage = statusMessage || null;
+            build.statusMessageType = statusMessageType || null;
+        }
+    }
```

Comment:
```suggestion
    // UNSTABLE -> SUCCESS needs to update meta and endtime.
    // However, the status itself cannot be updated to SUCCESS
    if (currentStatus === 'UNSTABLE') {
        return;
    }

    // Update the build status if desiredStatus is provided
    if (desiredStatus !== undefined) {
        build.status = desiredStatus;
    }

    // Handle different statuses and set the appropriate status message
    switch (build.status) {
        case 'ABORTED':
            build.statusMessage = currentStatus === 'FROZEN'
                ? `Frozen build aborted by ${username}`
                : `Aborted by ${username}`;
            break;
        case 'FAILURE':
        case 'SUCCESS':
            if (statusMessage) {
                build.statusMessage = statusMessage;
                build.statusMessageType = statusMessageType || null;
            }
            break;
        default:
            build.statusMessage = statusMessage || null;
            build.statusMessageType = statusMessageType || null;
            break;
    }
```
to improve readability: 
- use early return
- use switch instead of nested if else.

Reply:
Done